extensions/native/recursion/src/fri/two_adic_pcs.rs

```diff
@@ -224,69 +168,91 @@ pub fn verify_two_adic_pcs<C: Config>(
+                    let x = builder.get(&tag_exp, height_idx);
+                    builder.cycle_tracker_end("exp-reverse-bits-len");
+
+                    let hint_id = opened_values.id.clone();
+                    let ood_point_idx: Usize<C::N> = builder.eval(C::N::ZERO);
+                    iter_zip!(builder, mat_points, mat_values).for_each(|ptr_vec, builder| {
+                        let z: Ext<C::F, C::EF> = builder.iter_ptr_get(&mat_points, ptr_vec[0]);
+                        let ps_at_z = builder.iter_ptr_get(&mat_values, ptr_vec[1]);
+
+                        builder.cycle_tracker_start("single-reduced-opening-eval");
+
+                        if builder.flags.static_only {
+                            let n: Ext<C::F, C::EF> = builder.constant(C::EF::ZERO);
+                            let width = ps_at_z.len().value();
+                            let mat_opening_vals = {
+                                let witness_refs = builder.get_witness_refs(hint_id.clone());
+                                let start = hint_offset.value();
+                                witness_refs[start..start + width].to_vec()
+                            };
+                            for (i, v) in mat_opening_vals.into_iter().enumerate() {
+                                builder.set_value(&mat_opening, i, v.into());
+                            }
+                            for (t, alpha_pow) in alpha_pows.iter().take(width).enumerate() {
+                                let p_at_x = builder.get(&mat_opening, t);
+                                let p_at_z = builder.get(&ps_at_z, t);
+                                builder.assign(&n, n + (p_at_z - p_at_x) * alpha_pow.clone());
                             }
+                            builder.assign(&cur_ro, cur_ro + n / (z - x) * cur_alpha_pow);
+                            builder
+                                .assign(&cur_alpha_pow, cur_alpha_pow * alpha_pows[width].clone());
+                        } else {
+                            let mat_ro = builder.fri_single_reduced_opening_eval(
+                                alpha,
+                                hint_id.get_var(),
+                                hint_offset.get_var(),
+                                ood_point_idx.get_var(),
+                                &mat_opening,
+                                &ps_at_z,
+                            );
+                            builder.assign(&cur_ro, cur_ro + (mat_ro * cur_alpha_pow / (z - x)));
+                            builder.assign(&cur_alpha_pow, cur_alpha_pow * mat_alpha_pow);
+                        }
+                        builder.assign(&ood_point_idx, ood_point_idx.clone() + RVar::one());
+                        builder.cycle_tracker_end("single-reduced-opening-eval");
+                    });
+                    builder.assign(&hint_offset, hint_offset.clone() + mat_opening.len());
```

Comment:
I feel like `hint_offset` shouldn't be needed: can't this be managed by the chip purely during execution/trace generation?